### PR TITLE
Tools for GraphClust pipeline

### PR DESCRIPTION
**Pull request by eteriSokhoyan**

### CODE REVIEW

File: tools/EteriTools/CMFinder/cmFinder.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
+import glob, os, sys
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.py]: This should probably not be stored under EteriTools :)

File: tools/EteriTools/CMFinder/cmFinder.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+path = sys.argv[3]
+
+gapCmd = ""
+gapVal = ""
+if len(sys.argv) > 4:
+    gapCmd = sys.argv[4]
+    gapVal = sys.argv[5]
+
+
+#sh("mkdir  CLUSTER")
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.py]: Can you clean this up please and use flake8 to find some errors and best-practices for all your python files.

File: tools/EteriTools/CMFinder/cmFinder.py [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+#sh("cd CLUSTER/")
+cmd = " cp -f " + model_tree_stk  +  " model.cmfinder.stk"
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.py]: use string additions like `"foo %s bar %s\n" % (one, two)`

File: tools/EteriTools/CMFinder/cmFinder.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
+import glob, os, sys
+import shutil
+from shutil import copyfile
+import subprocess
+from os import system
+
+def sh(script):
+    system("bash -c '%s'" % script)
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.py]: This should be subprocess, isn't it?

File: tools/EteriTools/CMFinder/cmFinder.xml [hunk 5]
```diff
@@ -0,0 +1,69 @@
+<tool id="cmFinder" name="CMFinder_v0" version="0.1.0" >
+
+  <requirements>
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.xml]: indentation, 4 spaces, instead of a TAB

File: tools/EteriTools/CMFinder/cmFinder.xml [hunk 6]
```diff
@@ -0,0 +1,69 @@
+<tool id="cmFinder" name="CMFinder_v0" version="0.1.0" >
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.xml]: The name should be meaningful and you should have a description.

File: tools/EteriTools/CMFinder/cmFinder.xml [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+                            <remove value="&apos;"/>
+                        </valid>
+                    </sanitizer>
+                </param>
+            </when>
+            <when value=""/>
+        </conditional>
+    </inputs>
+    <outputs>
+       <data name="model_cmfinder_stk" format="stockholm" label="model_cmfinder_stk" from_work_dir = "model.cmfinder.stk"/>
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.xml]: spaces around = are not needed.

File: tools/EteriTools/CMFinder/cmFinder.xml [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+            <output name="model_cmfinder_stk" file="model.cmfinder.stk"/>
+        </test>
+    </tests>
+    <help>
+        <![CDATA[
+
+]]>
+</help>
+<citations>
+
```
Comment (bgruening) [inline tools/EteriTools/CMFinder/cmFinder.xml]: Graphclust paper? ViennaRNA? infernal?

File: tools/EteriTools/CollectResults/evaluation.py [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+from os import system
+import re
+#import sklearn
+#from sklearn import metrics
+#from tabulate import tabulate
+
+
+
+def sh(script):
+    system("bash -c '%s'" % script)
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: subprocess

File: tools/EteriTools/CollectResults/evaluation.py [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+
+#dataNames = "test-data/FASTA/data.names"
+dataNames = "FASTA/data.names"
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: hardcoded?

File: tools/EteriTools/CollectResults/evaluation.py [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+    with open(singleFile, "r") as f:
+        for line in f.readlines():
+            uniqueId = line.split()[6] #was 8
+            print "unique id = ", uniqueId
+            clustNum = line.split()[1]
+            rnaClass, sep, tail = uniqueId.partition("_")
+            #classToClusterDict[rnaClass] = clustNum
+            listOfClasses.append(rnaClass)
+            listOfClusters.append(clustNum)
+            print "rnaClass = ", rnaClass
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: Can you use logging if you want to produce debug output or remove this in the final script.

File: tools/EteriTools/CollectResults/evaluation.py [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+            print "clustNum = ", clustNum
+
+
+            with open(dataNames, "r") as names:
+                for nameLine in names.readlines():
+                    fullUniqeId = nameLine.split()[3] #waas 6
+                    print "full uniqueId = ", fullUniqeId
+                    rnaClass, sep, tail = fullUniqeId.partition("_")
+
+                    n = 2
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: ?
This can be added directly to the calculation, isn't it?

File: tools/EteriTools/CollectResults/evaluation.py [hunk 13]
```diff
@@ -0,0 +1,114 @@
...
+
+print "blackList = ", blackList
+print "numberOfClusters = " ,numberOfClusters
+numberOfClusters += 1 ### 1 cluster for all unassigned seqs
+with open(dataNames, "r") as names:
+    for nameLine in names.readlines():
+        fullUniqeId = nameLine.split()[3] # was 6
+        rnaClass, sep, tail = fullUniqeId.partition("_")
+        print "fullUniqeId = ", fullUniqeId
+        n = 2
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: same as above

File: tools/EteriTools/CollectResults/evaluation.py [hunk 14]
```diff
@@ -0,0 +1,114 @@
...
+        if short_unique not in blackList:
+
+            listOfClasses.append(rnaClass)
+            #listOfClasses.append(fullUniqeId)
+            listOfClusters.append(str(numberOfClusters))
+            numberOfClusters += 1 ### separate cluster for all unassigned seqs
+
+toWrite = ""
+#table=[]
+for i in range(len(listOfClusters)):
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: `for i in listOfClusters:` ?

File: tools/EteriTools/CollectResults/evaluation.py [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+    toWrite += listOfClasses[i] + "\t" + listOfClusters[i] + '\n'
+    #table += [[listOfClasses[i],listOfClusters[i]]]
+
+#f = open('table.txt', 'w')
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/evaluation.py]: please cleanup the scripts a little bit.

File: tools/EteriTools/CollectResults/glob_report.xml [hunk 16]
```diff
@@ -0,0 +1,195 @@
+<tool id="glob_report" name="Report_Results" version="0.1">
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/glob_report.xml]: name and description needs to be meaningful ...

File: tools/EteriTools/CollectResults/glob_report.xml [hunk 17]
```diff
@@ -0,0 +1,195 @@
...
+        #set $inputFiles = $inputFiles[:-1]
+
+        #set $inputFilesTrees = ""
+
+        #for $mods in $model_tree_files:
+            #set $inputFilesTrees += str($mods)+','
+        #end for
+        #set $inputFilesTrees = $inputFilesTrees[:-1]
+
+		    perl '$__tool_directory__/glob_res.pl' '$inputFiles' $merge_cluster_ol $merge_overlap $min_cluster_size $cm_min_bitscore $cm_max_eval $cm_bitscore_sig $partition_type '$__tool_directory__' $cut_type '$inputFilesTrees'
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/glob_report.xml]: indentation...

File: tools/EteriTools/CollectResults/glob_report.xml [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+        <param name="merge_cluster_ol" type="float" value="0.66" size="5" label="merge_cluster_ol" help="">
+            <sanitizer>
+                <valid initial="string.printable">
+                    <remove value="&apos;"/>
+                </valid>
+            </sanitizer>
+        </param>
+
+        <param name="merge_overlap" type="float" value="0.51" size="5" label="merge_overlap" help="">
+            <sanitizer>
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/glob_report.xml]: Why do you need a sanatizer for float values?

File: tools/EteriTools/CollectResults/glob_report.xml [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+                    <remove value="&apos;"/>
+                </valid>
+            </sanitizer>
+        </param>
+
+
+    </inputs>
+    <outputs>
+
+    <data name="final_stats" format="txt" from_work_dir="RESULTS/cluster.final.stats" label="cluster.final.stats"  />
```
Comment (bgruening) [inline tools/EteriTools/CollectResults/glob_report.xml]: indentation

File: tools/EteriTools/GSPAN/fasta2shrep_gspan.xml [hunk 20]
```diff
@@ -0,0 +1,146 @@
+<tool id="gspan" name="fasta_to_gspan" version="0.1">
+
+  <requirements>
+      <requirement type="package" version="2.1.6">rnashapes</requirement>
```
Comment (bgruening) [inline tools/EteriTools/GSPAN/fasta2shrep_gspan.xml]: please clean this one up, with indentation and cleaned newlines

File: tools/EteriTools/GSPAN/fasta2shrep_gspan.xml [hunk 21]
```diff
@@ -0,0 +1,146 @@
...
+
+        <param type="data" name="dataFasta" format="fasta" />
+
+        <param name="i_stacks" truevalue="-stack" falsevalue="" checked="True"  type="boolean" label="Add stacking information to graphs." help="-stack"/>
+        <param name="i_abstr" truevalue="-abstr" falsevalue=""  type="boolean" label="Add abstract structure graphs to the single shrep graph instances." help="-abstr"/>
+        <param name="i_crop_unpaired_ends" truevalue="-cue" falsevalue=""  type="boolean" label="Add abstract structure graphs to the single shrep graph instances." help="-cue"/>
+
+        <param name="M" type="integer" value="5" size="7" label="Max number of shreps that should be taken per window." help="-M">
+					<sanitizer>
+						<valid initial="string.printable">
```
Comment (bgruening) [inline tools/EteriTools/GSPAN/fasta2shrep_gspan.xml]: santizer and type int?

File: tools/EteriTools/GSPAN/fasta2shrep_gspan.xml [hunk 22]
```diff
@@ -0,0 +1,146 @@
...
+							<remove value="&apos;"/>
+						</valid>
+					</sanitizer>
+				</param>
+
+        <param name="rel_energy_range" type="integer" value="20" size="7" label=" Relative energy range, i.e. percentage (%) of MFE energy (RNAshapes)" help="-c">
+          <sanitizer>
+            <valid initial="string.printable">
+              <remove value="&apos;"/>
```
Comment (bgruening) [inline tools/EteriTools/GSPAN/fasta2shrep_gspan.xml]: sanitizer?

File: tools/EteriTools/NSPDK/NSPDK_candidateClusters.xml [hunk 23]
```diff
@@ -0,0 +1,265 @@
+<tool id="NSPDK_candidateClust" name="NSPDK_candidateClusters" version="9.2">
+   <!--parallelism method="multi" split_inputs="input1" split_mode="to_size" split_size="10" shared_inputs="" merge_outputs="output1, output2"></parallelism-->
```
Comment (bgruening) [inline tools/EteriTools/NSPDK/NSPDK_candidateClusters.xml]: help, description ...

File: tools/EteriTools/Preprocessing/preprocessing.xml [hunk 24]
```diff
@@ -0,0 +1,111 @@
...
+    '$fastaFile'  $max_length $in_winShift $min_seq_length
+
+]]>
+	</command>
+    <inputs>
+        <param type="data" name="fastaFile" format="fasta" />
+
+        <param name="max_length" type="integer" value="10000" size="5" label="window size">
+          <sanitizer>
+               <valid initial="string.printable">
```
Comment (bgruening) [inline tools/EteriTools/Preprocessing/preprocessing.xml]: I don't think that you need all this sanitizer :)

File: tools/EteriTools/Preprocessing/preprocessing.xml [hunk 25]
```diff
@@ -0,0 +1,111 @@
...
+        <output name="data.fasta" file="FASTA/data.fasta"/>
+        <output name="data.map" file="FASTA/data.map" />
+        <output name="data.names" file="FASTA/data.names"/>
+        <output name="data.fasta.scan" file="FASTA/data.fasta.scan" />
+        <!--output name="data.lens" file="FASTA/data.lens" /-->
+        <!--output name="data.locations" file="FASTA/data.locations" /-->
+        <!--output name="data.fasta.all_frags" file="FASTA/data.fasta.all_frags" /-->
+        <!--output name="orig.fasta" file="FASTA/orig.fasta"/-->
+        <!--output name="fragments.splitMasked.list" file="FASTA/fragments.splitMasked.list" /-->
+        <!--output name="FASTA" file="FASTA.zip"/-->
```
Comment (bgruening) [inline tools/EteriTools/Preprocessing/preprocessing.xml]: FASTA is neeed?

File: tools/sklearn/clf_metrics.xml [hunk 26]
```diff
@@ -32,8 +32,13 @@ y_p = columns("$clf_metrics.infile2","$clf_metrics.col2")
 
 options = params["clf_metrics"]["options"]
```
Comment (bgruening) [inline tools/sklearn/clf_metrics.xml]: Can you manage to put these changes in a separate PR please, as well es the macro changes.

File: tools/sklearn/clf_metrics.xml [hunk 27]
```diff
@@ -63,50 +68,52 @@ with open("$outfile", 'w+') as out_file:
...
                 <expand macro="clf_inputs"/>
                 <section name="options" title="Advanced Options" expanded="False">
                     <expand macro="beta"/>
-                    <!--labels-->
-                    <!--sample_weight-->
+
                     <expand macro="pos_label" default_value="1"/>
                     <expand macro="average">
-                        <option value="binary" selected="true" help="Only report results for the class specified by pos_label. This is applicable only if targets (y_{true,pred}) are binary.">binary</option>
+                        <option value="binary" selected="true" >binary</option>
```
Comment (bgruening) [inline tools/sklearn/clf_metrics.xml]: Why did you removed the help? Was planemo lint complaining?

**Comment by eteriSokhoyan** [comment]
@bgruening thak you for such  detailed comments, i'll do the changes you suggested. Didn't want to reply to every comment)

File: tools/EteriTools/CollectResults/evaluation.py (hunk 10 above)
Comment (eteriSokhoyan) [inline tools/EteriTools/CollectResults/evaluation.py]: here this is hardcoded because i pass FASTA folder to the tool and there all the names are defined by myself in 'preprocessing' step.

File: tools/EteriTools/Preprocessing/preprocessing.xml (hunk 25 above)
Comment (eteriSokhoyan) [inline tools/EteriTools/Preprocessing/preprocessing.xml]: yes it's really needed, because otherwise i have to pass to many separate files to the tools. In some cases it's more useful like this.

**Comment by bgruening** [comment]
@eteriSokhoyan both missing conda packages are now online. So if everything goes well this PR should turn green at some point :)

**Comment by mmiladi** [comment]
@eteriSokhoyan @bgruening Since most of the perl script is GraphClust code, we need to have the proper license for Eteri's tools. GraphClust comes with GPUv3.

**Comment by mmiladi** [comment]
@eteriSokhoyan Please try to reduce empty lines. As a rule of thumb remove more than two consecutive empty lines.

**Comment by eteriSokhoyan** [comment]
@bgruening so i removed sklearn folder completely from my branch problems solved)
for GraphClust we don't need sklearn, it was just for evaluation of final clusters but we can continue without it.

**Comment by bgruening** [comment]
I would not call it solved, but it's a step ahead.
If you look at the test results, you actually see that no test is running. Can you please create such a file: https://github.com/bgruening/galaxytools/blob/master/tools/rna_tools/vienna_rna/.shed.yml

And modify it so it fits to your tools. After this file is included travis will start testing.

**Comment by mmiladi** [comment]
>If you look at the test results, you actually see that no test is running. 

Have GraphClust wrapper passed the test?

**Comment by eteriSokhoyan** [comment]
ok, will do.

**Comment by eteriSokhoyan** [comment]
ok, so i added the .yml file to Prepocessing  and this one passed. 
so i will add .yml files one by one to each tool.

**Comment by bgruening** [comment]
Jupp, looks good. Please use a more specific name like `graphclust_preprocessing`. These yaml file is used to create automatically an entry in the Galaxy Appstore and the name is the repository name. So it should be somehow meaningful.

**Comment by bgruening** [comment]
@eteriSokhoyan please let me know when all tools have a shed.yaml and I can review them again.

**Comment by eteriSokhoyan** [comment]
@bgruening .shed.yml files added for all tools, so you can review them.

**Comment by bgruening** [comment]
@eteriSokhoyan do you know why this PR deletes so many test files e.g.? tools/sklearn/test-data/prp_result08

**Comment by eteriSokhoyan** [comment]
i'm not sure, but with sklearn could it be because in my local branch i deleted sklearn folder completly? Does it make sense for you?

**Comment by mmiladi** [comment]
@bgruening This PR should only ideally only add GraphClust folder to the master branch and keeps the sklearn directory untouched(to avoid triggering lint on sklearn folder). Is there a way we merge only GraphClust folder into the master branch?

**Comment by bgruening** [comment]
Yeah, removing a folder is not good. I can not merge it in this way. @eteriSokhoyan can you create a new branch, new PR with just your new tools? And a separate PR with the sklearn one?

**Comment by mmiladi** [comment]
@bgruening Sorry for our limited git knowledge :) : Eteri has already pushed sklearn changes to her forked master branch. She needs a *clean* pull of upstream master branch in a new(?) master branch **without** merging with her modified master branch. We tried to do it but failed because the sklearn lint was triggered.
Another solution would be to have a second clean fork of upstream repository in Eteri's github and then copy  tools there. Is it possible to have two fork of same repository in Github?

**Comment by eteriSokhoyan** [comment]
@bgruening , @mmiladi i created a new clean PR with only GraphClust tools. I'm closing this PR.
